pxr/imaging/hdSt/materialXFilter.cpp

```diff
@@ -1426,96 +1438,109 @@ size_t _BuildEquivalentMaterialNetwork(
     return topoHash.GetCode();
 }
 
+bool
+HdSt_MaterialFilterTask::IsMaterialX(SdrRegistry* sdrRegistry) const
+{
+    const SdrShaderNodeConstPtr mtlxSdrNode =
+        sdrRegistry->GetShaderNodeByIdentifierAndType(
+            terminalNode->nodeTypeId,
+            _tokens->mtlx);
+
+    return !!mtlxSdrNode;
+}
 
 mx::ShaderPtr
 HdSt_ApplyMaterialXFilter(
-    HdMaterialNetwork2* hdNetwork,
+    HdSt_MaterialFilterTask* filterTask,
     SdfPath const& materialPath,
-    HdMaterialNode2 const& terminalNode,
-    SdfPath const& terminalNodePath,
     HdSt_MaterialParamVector* materialParams,
     HdStResourceRegistry *resourceRegistry)
 {
-    // Check if the Terminal is a MaterialX Node
     SdrRegistry &sdrRegistry = SdrRegistry::GetInstance();
-    const SdrShaderNodeConstPtr mtlxSdrNode = 
-        sdrRegistry.GetShaderNodeByIdentifierAndType(terminalNode.nodeTypeId, 
-                                                     _tokens->mtlx);
-    if (!mtlxSdrNode) {
+
+    if (!filterTask->IsMaterialX(&sdrRegistry)) {
         return nullptr;
     }
 
     TRACE_FUNCTION_SCOPE("ApplyMaterialXFilter: Found Mtlx Node.")
 
-    // Anonymize the network to make sure shader code does not depend
-    // on node names
-    HdAnnonNodePathMap annonNodePathMap;
-    HdMaterialNetwork2 annonNetwork;
-    auto topoHash = _BuildEquivalentMaterialNetwork(
-        *hdNetwork, &annonNetwork, &annonNodePathMap);
-    SdfPath anonTerminalNodePath = annonNodePathMap[terminalNodePath];
-
-    mx::ShaderPtr glslfxShader;
-    const TfToken materialTag(_GetMaterialTag(*hdNetwork, terminalNode));
-    const bool bindlessTexturesEnabled = 
-        resourceRegistry->GetHgi()->GetCapabilities()->IsSet(
-            HgiDeviceCapabilitiesBitsBindlessTextures);
-    const TfToken apiName = resourceRegistry->GetHgi()->GetAPIName();
-
-    // Use the Resource Registry to cache the generated MaterialX 
-    // glslfx Shader
-    Tf_HashState shaderHash;
-    TfHashAppend(shaderHash, topoHash);
-    TfHashAppend(shaderHash, materialTag);
-    HdInstance<mx::ShaderPtr> glslfxInstance =
-        resourceRegistry->RegisterMaterialXShader(shaderHash.GetCode());
-
-    if (glslfxInstance.IsFirstInstance()) {
+    HdSt_MaterialXGeneratorTask generatorTask(
```

Comment:
Creating the generator task on the stack in the case when parallel codegen is off.